1-Draft/RFC00XX-Module-Isolation.md

```diff
@@ -0,0 +1,251 @@
+- `ExecutionContext.ActiveReflectionContext` is set before invoking a module command,
+  and restored to the previous context afterwards.
+- `ExecutionContext.ActiveReflectionContext` is set before running a command or script block in a module context (`& $module <cmd-or-script>`),
+  and restored to the previous context after the invocation.
+
+### Update Type Resolution Logic
+
+Now that we have custom ALC's for modules, the type resolving logic needs to be changed:
+
+1. We should first look for the type within the `ExecutionContext.ActiveReflectionContext`,
+and then revert back to the default ALC.
+1. When reverting back to the default ALC,
+we should ignore the assemblies that are loaded (and thus already searched) from the active ALC.
+
+Since different versions of the same assemblies can be loaded into a custom ALC in PowerShell,
+we may have type identity issues within a session -- type `C` and `C'` are the same type in terms of the fully-qualified type name,
+but they are from two different assembly instances that are loaded into two load contexts.
+
+In order to avoid type identity issue in type resolution and have it work consistently,
+we need to do the following extra updates:
+
+1. Today, we have a `TypeCache` data structure that maps a `Tuple<ITypeName, TypeResolutionState>` to a `Type`.
+We will need to have separate `TypeCache` defined in the `PSModuleLoadContext` class,
+so every custom ALC instance has its own type cache.
+The existing `TypeCache` will become the type cache for the default ALC.
+
+1. Today, we have `ExecutionContext.AssemblyCache` that holds all assemblies that are explicitly loaded by PowerShell.
+Simiarly, we will need to have separate `AssemblyCache` defined in the `PSModuleLoadContext` class,
+so every custom ALC instance has its own `AssemblyCache`.
+The existing `AssemblyCache` will be for the default ALC exclusively.
+
+1. All the `ITypeName` implementation types supports type caching.
+The cached type of a `ITypeName` instance should be able to invalid the caching when `ExecutionContext.ActiveReflectionContext` changes.
+
+## Potential Problems
+
+### Module Interoperability
+
+Imagine the assemblies A and B are loaded into ALC-M (isolated module M).
+Assembly A exposes a cmdlet `Demo-Cmdlet` that has a parameter accepting an object of type `B::C`.
+
+Scenario 1:
+Let's say the default ALC doesn't contains the assembly B.
+Then PowerShell won't be able to resolve the type `[B::C]` from outside the module M (e.g. from global scope),
+because `[B::C]` can only be resolved within the ALC-M.
+
+Scenario 2:
+Let's say a different version of B was loaded into the default ALC,
+and the result assembly instance is `B'`.
+So, from the global scope,
+one can create an object of `B'::C'` and call `Demo-Cmdlet` with it.
+That call will result in a type casting exception with the confusing error like "cannot cast B::C to B::C".
```

Comment:
Is there a possibility of us updating the error message here? We could try to catch the conversion error and then look to see if both types exist in known PS ALCs

Reply:
I think it's possible. Maybe just need to update the conversion code in `LanguagePrimitive`.